File: backend/app/models/advisory.py
from sqlalchemy import Column, ForeignKey, DateTime, String, Text, Enum, JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import uuid
import enum
from app.database import Base


class AdvisoryType(str, enum.Enum):
    IRRIGATION = "irrigation"
    FERTILIZER = "fertilizer"
    PEST_CONTROL = "pest_control"
    SOWING = "sowing"
    HARVESTING = "harvesting"
    WEATHER_ALERT = "weather_alert"
    GENERAL = "general"


class Advisory(Base):
    __tablename__ = "advisories"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    crop_record_id = Column(String(36), ForeignKey("crop_records.id"), nullable=True)
    farmer_id = Column(String(36), ForeignKey("farmers.id"), nullable=False, index=True)
    advisory_type = Column(Enum(AdvisoryType), nullable=False)
    title_en = Column(String(300), nullable=False)
    title_hi = Column(String(300), nullable=True)
    title_ta = Column(String(300), nullable=True)
    body_en = Column(Text, nullable=False)
    body_hi = Column(Text, nullable=True)
    body_ta = Column(Text, nullable=True)
    voice_url_en = Column(String(500), nullable=True)
    voice_url_hi = Column(String(500), nullable=True)
    voice_url_ta = Column(String(500), nullable=True)
    priority = Column(String(10), default="normal")  # low, normal, high, urgent
    extra_data = Column(JSON, default={})
    is_read = Column(String(5), default="false")
    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)

    crop_record = relationship("CropRecord", back_populates="advisories")

File: mobile/lib/models/advisory.dart
class Advisory {
  final String id;
  final String advisoryType;
  final String titleEn;
  final String? titleHi;
  final String? titleTa;
  final String bodyEn;
  final String? bodyHi;
  final String? bodyTa;
  final String? voiceUrlEn;
  final String? voiceUrlHi;
  final String? voiceUrlTa;
  final String priority;
  final String isRead;
  final String createdAt;

  const Advisory({
    required this.id,
    required this.advisoryType,
    required this.titleEn,
    required this.bodyEn,
    required this.priority,
    required this.isRead,
    required this.createdAt,
    this.titleHi,
    this.titleTa,
    this.bodyHi,
    this.bodyTa,
    this.voiceUrlEn,
    this.voiceUrlHi,
    this.voiceUrlTa,
  });

  String getTitle(String lang) {
    if (lang == 'hi' && titleHi != null) return titleHi!;
    if (lang == 'ta' && titleTa != null) return titleTa!;
    return titleEn;
  }

  String getBody(String lang) {
    if (lang == 'hi' && bodyHi != null) return bodyHi!;
    if (lang == 'ta' && bodyTa != null) return bodyTa!;
    return bodyEn;
  }

  factory Advisory.fromJson(Map<String, dynamic> json) {
    return Advisory(
      id: json['id']?.toString() ?? '',
      advisoryType: json['advisory_type']?.toString() ?? 'general',
      titleEn: json['title_en']?.toString() ?? '',
      titleHi: json['title_hi']?.toString(),
      titleTa: json['title_ta']?.toString(),
      bodyEn: json['body_en']?.toString() ?? '',
      bodyHi: json['body_hi']?.toString(),
      bodyTa: json['body_ta']?.toString(),
      voiceUrlEn: json['voice_url_en']?.toString(),
      voiceUrlHi: json['voice_url_hi']?.toString(),
      voiceUrlTa: json['voice_url_ta']?.toString(),
      priority: json['priority']?.toString() ?? 'normal',
      isRead: json['is_read']?.toString() ?? 'false',
      createdAt: json['created_at']?.toString() ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'advisory_type': advisoryType,
    'title_en': titleEn,
    'body_en': bodyEn,
    'priority': priority,
    'is_read': isRead,
    'created_at': createdAt,
  };
}
